### DISCUSSION TITLE
Customizing "org.opencontainers.image.title" value on artifact attach

### DISCUSSION THREAD

**Discussion by vvv444** [oras-project/oras#1405] [Q&A]
Hi,

Is there a way to customize the value of `org.opencontainers.image.title` on invocation of `oras attach`?
Now it always stores the file name/path on the system where the command is being executed, which is not always the desired result.

For example, when I prepare the artifact as temporary file, I get:
```
"annotations": {
    "org.opencontainers.image.title": "C:\\Users\\myuser\\AppData\\Local\\Temp\\tmpj41pib.tmp"
}
```

**Comment by qweeah** [comment]
It's by-design: ORAS users can use `oras push` to push files and they get what they pushed via `oras pull`. So basically `org.opencontainers.image.title` represents the file name.

Can you give a clearer user scenario of changing the file name? Like what do you expected to pull from the artifact that contains a temporary file?

**Reply by polarathene** [reply]
Is this annotation mandatory? It does not appear to be opt-out with `oras attach`, yet other tools seem to publish artifacts without the annotation.

The docs page makes no mention about the implicit annotation behaviour to use the filename provided as metadata. This does not occur with other commands like `oras manifest push`, so I did not give much thought to it.

For context in my case I was attaching an attestation produced by `docker buildx build`.

**Reply by Wwwsylvia** [reply]
Hi @polarathene, we are going to document this behavior in the `oras-go` repo. Here is the PR: https://github.com/oras-project/oras-go/pull/917
Feel free to review it and leave any comments!

**Reply by polarathene** [reply]
> Feel free to review it and leave any comments!

Thanks! It looks well written and structured, that was very informative :)

It doesn't explain why this annotation is mandatory, just that it is used as part of the file store.
- I assume the annotation is an oras specific requirement? Or would other tooling leverage this information to?
- It seems it should still be possible to rename the artifact via CLI arg if the field is required, rather than use the file system associated name.
- Additionally. I've not inspected if all the file attributes like permissions and ownerships are being preserved as part of this operation (_although that's presumably the case for directories due to tar_), a common issue I see with projects using GH releases is the tar archived release assets have file ownership UID/GID values from CI rather than resetting them 😓 (_usually desirable to be `0:0`_)

---

Since I was pushing to a registry (GHCR), my rough understanding is despite both src/dest being considered _Content Stores_ with a `Target`/`GraphTarget` interface, when I push a local file to GHCR via `oras attach` I'm copying from the _File Store_ to the _Repository Store_? (_rather than just a file upload, it's store/interface oriented_)

I haven't gone through `oras` docs enough yet, but if that already supports downloading the artifact to disk as a file again but at a different filename, then the annotation is just a default and it should be possible to support opt-out / replacement of that via the CLI during the upload 🤔

---

Is the documentation only intended to be at the repo README, or will it also have some visibility at the user facing docs site?

**Reply by Wwwsylvia** [reply]
> I assume the annotation is an oras specific requirement? Or would other tooling leverage this information to?

The annotation is an ORAS-specific implementation and is not compatible with other tools.

> It seems it should still be possible to rename the artifact via CLI arg if the field is required, rather than use the file system associated name.

Currently specifying a different name is not supported in CLI. You may open an issue if you think this feature would be helpful.

> Additionally. I've not inspected if all the file attributes like permissions and ownerships are being preserved as part of this operation

The file attributes are not preserved for a single file, but they are preserved for directories

---

> when I push a local file to GHCR via oras attach I'm copying from the File Store to the Repository Store?

Yes, exactly😀

>  then the annotation is just a default and it should be possible to support opt-out / replacement of that via the CLI during the upload

It's actually supported by the underlying `oras-go` SDK, but not in CLI for simplicity. Maybe you can open an issue and share your scenario 
there?

---

> Is the documentation only intended to be at the repo README, or will it also have some visibility at the user facing docs site?

We are considering putting a link to the doc in the official website: https://oras.land/docs/client_libraries/go
